pylabrobot/resources/plate_adapter.py

```diff
@@ -0,0 +1,175 @@
+      >>> Cos96_plate_1 = Cos_96_Rd(name='Cos96_plate_1')
+      >>> Alpaqua_magnum_flx_1.assign_child_resource(Cos96_plate_1)
+  """
+
+  def __init__(
+    self,
+    name: str,
+    size_x: float, size_y: float, size_z: float,
+    dx: float, dy: float, dz: float,
+    adapter_hole_size_x: float,
+    adapter_hole_size_y: float,
+    site_pedestal_z: float,
+    adapter_hole_dx: float = 9.0,
+    adapter_hole_dy: float = 9.0,
+    category: Optional[str] = None,
+    model: Optional[str] = None
+    ):
+    if adapter_hole_dx not in {9.0, 4.5, 2.25}:
+      raise ValueError("adapter_hole_dx must be one of 9.0, 4.5, or 2.25")
+    if adapter_hole_dy not in {9.0, 4.5, 2.25}:
+      raise ValueError("adapter_hole_dy must be one of 9.0, 4.5, or 2.25")
+
+    super().__init__(name=name, size_x=size_x, size_y=size_y, size_z=size_z,
+      category=category or "plate_adapter", model=model)
+
+    self._child_resource_location = None
+    self._child_resource: Optional[Resource] = None
+    self.dx = dx
+    self.dy = dy
+    self.dz = dz
+    self.adapter_hole_size_x = adapter_hole_size_x
+    self.adapter_hole_size_y = adapter_hole_size_y
+    self.adapter_hole_size_z = self.get_size_z() - self.dz
+    self.adapter_hole_dx = adapter_hole_dx
+    self.adapter_hole_dy = adapter_hole_dy
+    self.site_pedestal_z = site_pedestal_z
+
+  @property
+  def child_resource_location(self) -> Optional[Coordinate]:
+    return self._child_resource_location
+
+  def assign_child_resource(
+    self,
+    resource: Plate,
+    location: Optional[Coordinate] = None,
+    reassign: bool = True
+  ):
+    """Assign a Plate to a PlateAdapter. If `location` is not provided,
+    the resource will autoadjust the placement location based on the
+    PlateAdapter-Plate relationship."""
+
+    if self._child_resource is not None and not reassign:
+      raise ValueError(f"{self.name} already has a child resource assigned")
+    if not isinstance(resource, Plate):
+      raise ValueError("Only plates can be assigned to Alpaqua 96 magnum flx.")
+
+    # TODO: have discussion oon whether to transfer flat bottom error checking
+    # TODO: check whether allPlate children information could
+    # be made accessible from the Plate class
+
+    # Calculate Plate information (which is not directly accessible from the Plate class)
+    x_locations = sorted(OrderedDict.fromkeys([well_n.location.x
+      for well_n in resource.children]))
+    y_locations = sorted(OrderedDict.fromkeys([well_n.location.y
+      for well_n in resource.children]))
+
+    def calculate_well_spacing(float_list: List[float]):
+      """ Calculate the difference between every x and x+1 element in the list of floats. """
+      if len(float_list) < 2:
+        return None
+      differences = [round(float_list[i+1] - float_list[i],2) for i in range(len(float_list) - 1)]
+      if differences[0] is None:
+        raise ValueError("well spacing has to be uniform, and cannot be None")
```

Comment:
how could this be None?

Reply:
I think it would crash on line 137.

Reply:
> how could this be None?

This is a workaround that came  about in the type checking process:
```python
x_locations = sorted(OrderedDict.fromkeys([well_n.location.x
      for well_n in resource.children]))
    y_locations = sorted(OrderedDict.fromkeys([well_n.location.y
      for well_n in resource.children]))
```

If I remember correctly, this generated `None` if the adapter doesn't have a Plate on it, in the adapter assignment command.

Reply:
i am slightly confused by why OrderedDict is used here.

what I meant:
```python
differences = [round(float_list[i+1] - float_list[i],2) for i in range(len(float_list) - 1)]
if differences[0] is None:
```

how can `round(float_list[i+1] - float_list[i],2)` return None?

Reply:
> i am slightly confused by why OrderedDict is used here.

ahhh that one I can answer easily: I needed a `np.unique()` without having to make PLR dependent on numpy, OrderedDict is part of the standard library `collections`, I am happy to change it to a viable alternative though

> how can `round(float_list[i+1] - float_list[i],2)` return None?
Tbh I cannot remember. I agree that it should never come to this. We can remove it and see whether anything breaks?

Reply:
hb a [`set`](https://docs.python.org/3/tutorial/datastructures.html#sets)? :)

> We can remove it and see whether anything breaks?
sounds good

Reply:
went ahead with https://github.com/PyLabRobot/pylabrobot/pull/152/commits/5d3b8208728e39fb3aedfd6fb186e3dc6abbee1f